src/internal/fileserver/fileserver.go

```diff
@@ -0,0 +1,415 @@
+		w.Header().Set("vary", constants.ContextTokenKey)
+		ourRequest.get(ctx)
+		return
+	default:
+		ourRequest.displayErrorf(ctx, http.StatusMethodNotAllowed, "unknown HTTP method %q", req.Method)
+		return
+	}
+}
+
+func (r *Request) get(ctx context.Context) {
+	req := r.Request
+	parts := strings.Split(req.URL.Path, "/")
+	switch {
+	case len(parts) == 5 && parts[4] != "":
+		// If parts[4] is "", then they are requesting what looks like a "list commits"
+		// page, that we don't implement.  Otherwise, they want the listing of the root
+		// directory, but forgot the trailing /.
+		http.Redirect(r.ResponseWriter, r.Request, r.Request.URL.Path+"/", http.StatusMovedPermanently)
+		return
+	case len(parts) >= 6: // /pfs/project/repo/commit|branch/path/to/some/file
+		parts[5] = path.Join(parts[5:]...)
+	default:
+		// Someday we can have a projects page, project repo page, repo commit page, etc.
+		// For now, only files and directories.
+		r.displayErrorf(ctx, http.StatusNotFound,
+			//                     0 1   2         3      4               5
+			"invalid URL; expecting /pfs/<project>/<repo>/<commit|branch>/<path...>, got %v",
+			strings.Join(parts, "/"))
+		return
+	}
+	if got, want := parts[1], "pfs"; got != want {
+		r.displayErrorf(ctx, http.StatusInternalServerError, "unexpectedly handling request not for /pfs; got %v want %v", got, want)
+		return
+	}
+	file := &pfs.File{
+		Path: parts[5],
+	}
+	repo := &pfs.Repo{
+		Name: parts[3],
+		Type: pfs.UserRepoType,
+		Project: &pfs.Project{
+			Name: parts[2],
+		},
+	}
+	if commitish := parts[4]; uuid.IsUUIDWithoutDashes(commitish) {
+		file.Commit = &pfs.Commit{
+			Repo: repo,
+			Id:   commitish,
+		}
+		var finished bool
+		if commit, err := r.PachClient.PfsAPIClient.InspectCommit(ctx, &pfs.InspectCommitRequest{
+			Commit: file.Commit,
+		}); err == nil {
+			finished = commit.GetFinished() != nil
+		}
+		if finished {
+			// If the commit is finished, the user can cache this content
+			// forever.
+			r.ResponseWriter.Header().Set("cache-control", "private")
+		} else {
+			// Open commits can change, so they aren't safe to cache.  Note that
+			// "no-cache" actually means that the response can be cached, but that the
+			// client has to revalidate (with If-None-Match etc.) before using the
+			// cached result.  That's exactly what we want.
+			r.ResponseWriter.Header().Set("cache-control", "private, no-cache")
+		}
+	} else {
+		file.Commit = &pfs.Commit{
+			Branch: &pfs.Branch{
+				Repo: repo,
+				Name: commitish,
+			},
+		}
+		// Branch references are never cacheable; the branch can move at any time.
+		r.ResponseWriter.Header().Set("cache-control", "private, no-cache")
+	}
+	info, err := r.PachClient.PfsAPIClient.InspectFile(ctx, &pfs.InspectFileRequest{
+		File: file,
+	})
+	if err != nil {
+		r.displayGRPCError(ctx, "problem inspecting file", err)
+		return
+	}
+
+	// Set headers for future conditional requests.
+	etag := `"` + hex.EncodeToString(info.GetHash()) + `"`
```

Comment:
Fun fact: Spark does not quote Etag header values in its requests, and actual S3 doesn't care if they're quoted or not, but the HTTP spec requires them to be quoted for no apparent reason, and Go's HTTP server insta-fails any request with unquoted etag header values. Really makes you love working with computers.